azure-quantum/azure/quantum/workspace.py

```diff
@@ -374,8 +374,18 @@ def login(self, refresh:bool=False) -> Authentication:
             the user credentials to authenticate with Azure Quantum
         """
 
-        if refresh or (self.credentials is None):
-            msal_wrapper = MsalWrapper(self.subscription_id, refresh=refresh)
-            self.credentials = msal_wrapper.acquire_auth_token()
+        if refresh:

```

Comment:
It looks to me that the "refresh" param is incompatible with ServicePrincipalCredentials flow. If a customer sets credentials field to a ServicePrincipalCredentials and then sets refresh = True, we would wipe out the credentials and try acquiring a authToken using the user-based token flow. Not suggesting a change here but something to reconcile in a future change.

Reply:
You are right. I updated the logic to not remove the ServicePrincipalCredentials that the customer could have added.
I think we can improve it more, but this should be good enough for now.